Add a guideline for mixins (#9016)
mdn/browser-compat-data: docs/data-guidelines.md

If you're adding new data for Node.js, use `0.10.0` or later. If a feature was added in a version before `0.10.0`, use `0.10.0` anyway.

This guideline was proposed in [#6861](https://github.com/mdn/browser-compat-data/pull/6861).

## Mixins

> Mixins appear in WebIDL and in specifications to define Web APIs. For web developers, they aren't observable directly; they act as helpers to avoid repeating API definitions. BCD doesn't expose mixins.

Reviewer: Well, we _do_ expose some mixins. Maybe it'll help if we give an imperative. Something like…

```suggestion
Mixins appear in WebIDL and in specifications to define Web APIs. For web developers, they aren't observable directly; they act as helpers to avoid repeating API definitions. Don't add mixins to BCD where they do not already exist.
```

> Mixins appear in WebIDL and in specifications to define Web APIs. For web developers, they aren't observable directly; they act as helpers to avoid repeating API definitions. Don't add mixins to BCD where they do not already exist.

Reviewer 2: ```suggestion
[Interface mixins](https://heycam.github.io/webidl/#idl-interface-mixins) in Web IDL are used in specifications to define Web APIs. For web developers, they aren't observable directly; they act as helpers to avoid repeating API definitions. Don't add mixins to BCD where they do not already exist.
```

For example, [`HTMLHyperlinkElementUtils`](https://html.spec.whatwg.org/multipage/links.html#htmlhyperlinkelementutils) is a mixin defined in the HTML specification.

> Members of this mixin are available to `HTMLAnchorElement` and `HTMLAreaElement`, that's where BCD exposes them, and thus there are two options to add mixin members to BCD:

Reviewer: (under the assumption that #8933 merges promptly—maybe we make sure they merge together)

```suggestion
Members of this mixin are available to `HTMLAnchorElement` and `HTMLAreaElement`, that's where BCD exposes them. Add mixin members to BCD in one of these ways:
```

> 1. Members of `HTMLHyperlinkElementUtils` are added directly to the `api/HTMLAnchorElement.json` and `api/HTMLAreaElement.json` files as if they were regular members of these interfaces.

Reviewer: ```suggestion
1. For smaller mixins, add members of `HTMLHyperlinkElementUtils` directly to the `api/HTMLAnchorElement.json` and `api/HTMLAreaElement.json` files as if they were regular members of these interfaces.
```

> 2. (Large) Mixins have their own files in the `api/_mixins/` folder and indicate for which interface they are using file names like: `HTMLHyperlinkElementUtils__HTMLAnchorElement.json` and `HTMLHyperlinkElementUtils__HTMLAreaElement.json`.  

Reviewer: ```suggestion
2. For larger mixins, create a file in the `api/_mixins/` folder and indicate for which interface they are using file names like: `HTMLHyperlinkElementUtils__HTMLAnchorElement.json` and `HTMLHyperlinkElementUtils__HTMLAreaElement.json`.  
```

> 2. For larger mixins, create a file in the `api/_mixins/` folder and indicate for which interface they are using file names like: `HTMLHyperlinkElementUtils__HTMLAnchorElement.json` and `HTMLHyperlinkElementUtils__HTMLAreaElement.json`.  

Reviewer 2: Any particular reason there should be two underscores in the filenames? Is it because some Web IDL interface names have underscores in them? I'd just ignore that, or maybe use a separator that can't appear in interface/mixin names, like a dash.

Author: Originally, I proposed `HTMLHyperlinkElementUtils-for-HTMLAnchorElement.json`, but `HTMLHyperlinkElementUtils-HTMLAnchorElement.json` would be fine with me, too. @ddbeck any reason you proposed two underscores?

Reviewer: I just thought it made the IDL part stand out from the the non-IDL part better, as compared to one. I was just throwing an idea out there. I don't feel strongly about how it's done.

Author: I guess two underscores is fine then. It does separate them nicely.

>    Note that in these files you'll need to expose the data under the correct tree, so for `HTMLHyperlinkElementUtils__HTMLAnchorElement.json` the file needs to start like this:

Reviewer: ```suggestion
   In these files, expose the data under the correct tree. For `HTMLHyperlinkElementUtils__HTMLAnchorElement.json`, the file needs to start like this:
```

> ```json

Reviewer: The tiniest nit ever: I'm not sure why the diff paints this red. Maybe because the JSON is invalid? I looked at the rendered file on your branch directly and it seems fine, though that caused me to notice that this is not indented to the level of the `2.` body text. you could add three spaces to each of the code block lines to make it align to the parent list item.

Otherwise, marking this as approved, with the expectation that we'll merge it at the same time as the other PR. 👍

Reviewer 2: I think it's red because the JSON isn't terminated. It's the same style as you get for invalid JSON, often if you put `...` in an example.

Author: I've fixed the indentation and I've removed the json highlighting, so that it doesn't try to process it.
